File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/wordlist/presentation/WordsPresenter.java
package com.przemyslawlusnia.vocabularycreator.wordlist.presentation;

import com.przemyslawlusnia.vocabularycreator.core.BasePresenter;
import com.przemyslawlusnia.vocabularycreator.core.RxUseCase;
import com.przemyslawlusnia.vocabularycreator.wordlist.domain.AddWordUseCase;
import com.przemyslawlusnia.vocabularycreator.wordlist.domain.DeleteWordUseCase;
import com.przemyslawlusnia.vocabularycreator.wordlist.domain.WordDomainModel;
import java.util.List;
import rx.Observer;
import rx.Subscription;
import rx.subscriptions.Subscriptions;

public class WordsPresenter extends BasePresenter<WordsView> {

  private static final String TAG = WordsPresenter.class.getSimpleName();

  private final WordViewMapper mapper;
  private final AddWordUseCase addWordUseCase;
  private final DeleteWordUseCase deleteWordUseCase;
  private final RxUseCase<List<WordDomainModel>> getAllWordsUseCase;
  private final rx.Scheduler uiScheduler;
  private Subscription addWordSubscription = Subscriptions.empty();
  private Subscription deleteWordSubscription = Subscriptions.empty();
  private Subscription getAllWordsSubscription = Subscriptions.empty();

  public WordsPresenter(WordsView wordsView,
                        WordViewMapper mapper,
                        rx.Scheduler uiScheduler,
                        AddWordUseCase addWordUseCase,
                        DeleteWordUseCase deleteWordUseCase,
                        RxUseCase<List<WordDomainModel>> getAllWordsUseCase) {
    this.view = wordsView;
    this.mapper = mapper;
    this.uiScheduler = uiScheduler;
    this.addWordUseCase = addWordUseCase;
    this.deleteWordUseCase = deleteWordUseCase;
    this.getAllWordsUseCase = getAllWordsUseCase;
  }

  public void addWord(WordViewModel wordViewModel) {
    view.showProgress();
    addWordUseCase.init(mapper.mapToWordDomainModel(wordViewModel));
    addWordSubscription = addWordUseCase.execute()
        .observeOn(uiScheduler)
        .subscribe(new AddWordObserver(wordViewModel));
  }

  public void deleteWords(List<WordViewModel> wordViewModels) {
    view.showProgress();
    deleteWordUseCase.init(mapper.mapToWordDomainModels(wordViewModels));
    deleteWordSubscription = deleteWordUseCase.execute()
        .observeOn(uiScheduler)
        .subscribe(new DeleteWordObserver());
  }

  public void getAllWords() {
    view.showProgress();
    getAllWordsSubscription = getAllWordsUseCase.execute()
        .map(wordDomainModels -> mapper.mapToWordViewModels(wordDomainModels))
        .observeOn(uiScheduler)
        .subscribe(new GetAllWordsObserver());
  }

  @Override
  public void onUnsubscribe() {
    if (!deleteWordSubscription.isUnsubscribed()) {
      deleteWordSubscription.unsubscribe();
    }
    if (!getAllWordsSubscription.isUnsubscribed()) {
      getAllWordsSubscription.unsubscribe();
    }
    if (!addWordSubscription.isUnsubscribed()) {
      addWordSubscription.unsubscribe();
    }
  }

  // Observers

  private class AddWordObserver implements Observer<Boolean> {
    private final WordViewModel wordViewModel;

    public AddWordObserver(WordViewModel wordViewModel) {
      this.wordViewModel = wordViewModel;
    }

    @Override
    public void onCompleted() {
      view.hideProgress();
    }

    @Override
    public void onError(Throwable t) {
      view.hideProgress();
      view.showFailure(t.getMessage());
    }

    @Override
    public void onNext(Boolean success) {
      if (success) {
        view.showAddWord(wordViewModel);
      }
    }
  }

  private class GetAllWordsObserver implements Observer<List<WordViewModel>> {

    @Override
    public void onCompleted() {
      view.hideProgress();
    }

    @Override
    public void onError(Throwable t) {
      view.hideProgress();
      view.showFailure(t.getMessage());
    }

    @Override
    public void onNext(List<WordViewModel> wordViewModels) {
      view.showAllWords(wordViewModels);
    }

  }

  private class DeleteWordObserver implements Observer<Void> {

    @Override
    public void onCompleted() {
      view.showDeletedWords();
      view.hideProgress();
    }

    @Override
    public void onError(Throwable t) {
      view.hideProgress();
      view.showFailure(t.getMessage());
    }

    @Override
    public void onNext(Void aVoid) {
      // nothing
    }
  }
}

File: app/src/main/java/com/przemyslawlusnia/vocabularycreator/wordlist/presentation/WordsPresenter.kt
package com.przemyslawlusnia.vocabularycreator.wordlist.presentation

import com.przemyslawlusnia.vocabularycreator.core.BasePresenter
import com.przemyslawlusnia.vocabularycreator.core.RxUseCase
import com.przemyslawlusnia.vocabularycreator.wordlist.domain.AddWordUseCase
import com.przemyslawlusnia.vocabularycreator.wordlist.domain.DeleteWordUseCase
import com.przemyslawlusnia.vocabularycreator.wordlist.domain.WordDomainModel
import rx.Observer
import rx.subscriptions.Subscriptions

class WordsPresenter(private val mapper: WordViewMapper,
                     private val uiScheduler: rx.Scheduler,
                     private val addWordUseCase: AddWordUseCase,
                     private val deleteWordUseCase: DeleteWordUseCase,
                     private val getAllWordsUseCase: RxUseCase<List<WordDomainModel>>) : BasePresenter<WordsView>() {
    private var addWordSubscription = Subscriptions.empty()
    private var deleteWordSubscription = Subscriptions.empty()
    private var getAllWordsSubscription = Subscriptions.empty()

    fun addWord(wordViewModel: WordViewModel) {
        view?.showProgress()
        addWordUseCase.init(mapper.mapToWordDomainModel(wordViewModel))
        addWordSubscription = addWordUseCase.execute()
                .observeOn(uiScheduler)
                .subscribe(AddWordObserver(wordViewModel))
    }

    fun deleteWords(wordViewModels: List<WordViewModel>) {
        view?.showProgress()
        deleteWordUseCase.init(mapper.mapToWordDomainModels(wordViewModels))
        deleteWordSubscription = deleteWordUseCase.execute()
                .observeOn(uiScheduler)
                .subscribe(DeleteWordObserver())
    }

    fun getAllWords() {
        view?.showProgress()
        getAllWordsSubscription = getAllWordsUseCase.execute()
                .map { wordDomainModels -> mapper.mapToWordViewModels(wordDomainModels) }
                .observeOn(uiScheduler)
                .subscribe(GetAllWordsObserver())
    }

    override fun onUnsubscribe() {
        if (!deleteWordSubscription.isUnsubscribed) {
            deleteWordSubscription.unsubscribe()
        }
        if (!getAllWordsSubscription.isUnsubscribed) {
            getAllWordsSubscription.unsubscribe()
        }
        if (!addWordSubscription.isUnsubscribed) {
            addWordSubscription.unsubscribe()
        }
    }

    // Observers

    private inner class AddWordObserver(private val wordViewModel: WordViewModel) : Observer<Boolean> {

        override fun onCompleted() {
            view?.hideProgress()
        }

        override fun onError(t: Throwable) {
            view?.hideProgress()
            view?.showFailure(t.message)
        }

        override fun onNext(success: Boolean?) {
            if (success!!) {
                view?.showAddWord(wordViewModel)
            }
        }
    }

    private inner class GetAllWordsObserver : Observer<List<WordViewModel>> {

        override fun onCompleted() {
            view?.hideProgress()
        }

        override fun onError(t: Throwable) {
            view?.hideProgress()
            view?.showFailure(t.message)
        }

        override fun onNext(wordViewModels: List<WordViewModel>) {
            view?.showAllWords(wordViewModels)
        }

    }

    private inner class DeleteWordObserver : Observer<Void> {

        override fun onCompleted() {
            view?.showAllWordsAfterDeletion()
            view?.hideProgress()
        }

        override fun onError(t: Throwable) {
            view?.hideProgress()
            view?.showFailure(t.message)
        }

        override fun onNext(aVoid: Void) {
            // nothing
        }
    }

}
